wallabagger/js/background.js

```diff
@@ -1,207 +1,457 @@
+                        })
+                        .catch(a => {
+                            api.clear();
+                            // api.save();
+                            postIfConnected({ response: 'setup-checkurl', data: api.data, result: false });
+                        });
+                break;
+            case 'deleteArticleTag':
+                if (msg.articleId !== -1) {
+                    api.DeleteArticleTag(msg.articleId, msg.tagId).then(data => {
+                        postIfConnected({ response: 'articleTags', tags: data.tags });
+                        cache.set(msg.tabUrl, data);
+                    });
+                } else {
+                    dirtyCacheSet(msg.tabUrl, {tagList: msg.tags});
+                }
+                break;
+            case 'saveTags':
+                if (msg.articleId !== -1) {
+                    api.SaveTags(msg.articleId, msg.tags).then(data => {
+                        postIfConnected({ response: 'articleTags', tags: data.tags });
+                        cache.set(msg.tabUrl, data);
+                    });
+                } else {
+                    dirtyCacheSet(msg.tabUrl, {tagList: msg.tags});
+                }
+                break;
+            case 'SaveStarred':
+            case 'SaveArchived':
+                if (msg.articleId !== -1) {
+                    api[msg.request](msg.articleId, msg.value ? 1 : 0).then(data => {
+                        postIfConnected({ response: 'action', value: {starred: data.is_starred === 1, archived: data.is_archived === 1} });
+                        cache.set(msg.tabUrl, data);
+                    });
+                } else {
+                    dirtyCacheSet(msg.tabUrl, (msg.request === 'SaveStarred') ? {is_starred: msg.value ? 1 : 0} : {is_archived: msg.value ? 1 : 0});
+                }
+                break;
+            default: {
+                console.log(`unknown request ${JSON.stringify(msg)}`);
+            }
+        }
+    } catch (error) {
+        setIcon(icons.bad);
+        setTimeout(function () { setIcon(icons.default); }, 5000);
+        postIfConnected({ response: 'error', error: error });
+    }
+}
 
-        browser.tabs.onCreated.addListener(function (tab) {
-            browser.browserAction.setIcon({ path: icon.default });
-        });
+function onRuntimeConnect (port) {
+    Port = port;
+    portConnected = true;
 
-        browser.tabs.onUpdated.addListener(function (tabId, changeInfo, tab) {
-            if (changeInfo.status === 'loading' && tab.active) {
-                requestExists(tab.url);
-            }
-        });
+    Port.onDisconnect.addListener(function () { portConnected = false; });
+    Port.onMessage.addListener(onPortMessage);
+}
+
+function addListeners () {
+    browser.contextMenus.onClicked.addListener(onContextMenusClicked);
+    browser.commands.onCommand.addListener(onCommandsCommand);
```

Comment:
_execute_browser_action is handled by browser itself
we have another shortcut for background saving - it have its own command "wallabag-it" - that we have to handle separately.